test/unit/math/prim/prob/hmm_marginal_test.cpp

```diff
@@ -0,0 +1,269 @@
+  // the columns sum to 1. This allows us to do finite diff tests,
+  // without violating the simplex constraint of rho and Gamma.
+  std::vector<double> rho_unconstrained;
+  Eigen::MatrixXd Gamma_unconstrained;
+  stan::test::ad_tolerances tols;
+};
+
+// TEST(hmm_marginal_lpdf, two_state) {
+TEST_F(hmm_marginal_lpdf_test, ten_transitions) {
+  using stan::math::hmm_marginal_lpdf;
+
+  // CHECK -- EXPECT_EQ returns an error.
+  EXPECT_FLOAT_EQ(-18.37417, hmm_marginal_lpdf(log_omegas, Gamma, rho));
+
+  // Differentiation tests
+  auto hmm_functor = [](const auto& log_omegas, const auto& Gamma_unconstrained,
+                        const auto& rho_unconstrained) {
+    return hmm_marginal_test_wrapper(log_omegas, Gamma_unconstrained,
+                                     rho_unconstrained);
+  };
+
+  stan::test::expect_ad(tols, hmm_functor, log_omegas, Gamma_unconstrained,
+                        rho_unconstrained);
+}
+
+TEST_F(hmm_marginal_lpdf_test, zero_transitions) {
+  using stan::math::hmm_marginal_lpdf;
+
+  EXPECT_FLOAT_EQ(-1.520827, hmm_marginal_lpdf(log_omegas_zero, Gamma, rho));
+
+  // Differentiation tests
+  auto hmm_functor = [](const auto& log_omegas, const auto& Gamma_unconstrained,
+                        const auto& rho_unconstrained) {
+    return hmm_marginal_test_wrapper(log_omegas, Gamma_unconstrained,
+                                     rho_unconstrained);
+  };
+
+  stan::test::expect_ad(tols, hmm_functor, log_omegas_zero, Gamma_unconstrained,
+                        rho_unconstrained);
+}
+
+TEST(hmm_marginal_lpdf, one_state) {
+  using stan::math::hmm_marginal_lpdf;
+  int n_states = 1, p1_init = 1, gamma1 = 1, n_transitions = 10, abs_mu = 1,
+      sigma = 1;
+  Eigen::VectorXd rho(n_states);
+  rho << p1_init;
+  Eigen::MatrixXd Gamma(n_states, n_states);
+  Gamma << gamma1;
+  Eigen::VectorXd obs_data(n_transitions + 1);
+  obs_data << -0.9692032, 1.6367754, 1.0339449, 0.9798393, 0.4829358, 2.7508704,
+      0.3122448, 1.8316583, 1.6327319, 1.2097332, 0.4087620;
+  Eigen::MatrixXd log_omegas(n_states, n_transitions + 1);
+  for (int n = 0; n < n_transitions + 1; n++)
+    log_omegas.col(n)[0] = state_lpdf(obs_data[n], abs_mu, sigma, 0);
+
+  EXPECT_FLOAT_EQ(-14.89646, hmm_marginal_lpdf(log_omegas, Gamma, rho));
+
+  // Differentiation tests
+  std::vector<double> rho_unconstrained(n_states - 1);
+  for (int i = 0; i < rho.size(); i++)
+    rho_unconstrained[i] = rho(i);
+  Eigen::MatrixXd Gamma_unconstrained
+      = Gamma.block(0, 0, n_states, n_states - 1);
+
+  auto hmm_functor = [](const auto& log_omegas, const auto& Gamma_unconstrained,
+                        const auto& rho_unconstrained) {
+    return hmm_marginal_test_wrapper(log_omegas, Gamma_unconstrained,
+                                     rho_unconstrained);
+  };
+
+  stan::test::ad_tolerances tols;
+  double infinity = std::numeric_limits<double>::infinity();
+  tols.hessian_val_ = infinity;
```

Comment:
Does this not work for forward mode?  If not, why not?  It may just be a matter of using the right intermediate types in operands-and-partials.

Reply:
The differentiation algorithm is constructed as an adjoint method, but since it directly computes a log density, it could in principle work with fwd diff. I relaxed some of the constraints and got the following to work:
```
    // tols.hessian_val_ = infinity;
    // tols.hessian_grad_ = infinity;
    tols.hessian_hessian_ = infinity;
    // tols.hessian_fvar_val_ = infinity;
    tols.hessian_fvar_grad_ = infinity;
    tols.hessian_fvar_hessian_ = infinity;
    // tols.grad_hessian_val_ = infinity;
    // tols.grad_hessian_hessian_ = infinity;
    tols.grad_hessian_grad_hessian_ = infinity;
```
It looks like fwd mode is indeed not working. I'm not sure what the right intermediate types would need to be for operands-and-partials, my use of them seemed pretty standard.

Reply:
Could this be because you use `double` and `value_of_rec`? Since `double` would be for `var`'s inner type and value_of_rec returns back the double value for a nested fvar. Maybe try replacing the places you use double with `partials_type_t<Stuff>` and use `value_of` instead of `value_of_rec`

Reply:
There are a lot of internal operations that use double types. How can I template them appropriately? I'm not sure how to infer the fvar's inner type from the argument.

Reply:
> There are a lot of internal operations that use double types. How can I template them appropriately? 

What specifically? Pretty much everything should have a generic method in prim unless it has a specialization in rev/fwd

> I'm not sure how to infer the fvar's inner type from the argument.

`partial_type_t` should do this since partial_type for fvar is defined as

```cpp
template <typename T>
struct partials_type<T, require_fvar_t<T>> {
  using type = typename std::decay_t<T>::Scalar;
};
```